Implemented (field_n,field) flow formulation

francoishamon opened the pull request:
This PR implements the (`field_n`,`field`) flow formulation as discussed during the developer meeting, and removes all the code that was previously doing `field+deltaField`.

I get numerical diffs for the proppant transport examples because:
-  in `develop` the fluid update was using `proppantConcentration + dProppantConcentration` [here](https://github.com/GEOSX/GEOSX/blob/8445c1c2a70eb0bf42a5c25f9c079f71fd16617e/src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseProppantBaseKernels.hpp#L50), 
- but this call was happening after incrementing `proppantConcentration` in `implicitStepComplete` [here](https://github.com/GEOSX/GEOSX/blob/8445c1c2a70eb0bf42a5c25f9c079f71fd16617e/src/coreComponents/physicsSolvers/fluidFlow/proppantTransport/ProppantTransport.cpp#L456), 
- and before zeroing out ` dProppantConcentration` in `resetStateToBeginningOfStep` [here](https://github.com/GEOSX/GEOSX/blob/8445c1c2a70eb0bf42a5c25f9c079f71fd16617e/src/coreComponents/physicsSolvers/fluidFlow/proppantTransport/ProppantTransport.cpp#L951),
- so the fluid properties were computed with a `proppantConcentration` incremented twice. 

For other integrated tests, I don't get numerical diffs.

The PR is ready for merge.

Solves #1875 

Solves #1883

Integrated tests rebaselined in https://github.com/GEOSX/integratedTests/pull/216

klevzoff commented on src/coreComponents/physicsSolvers/fluidFlow/wells/CompositionalMultiphaseWell.cpp:
```diff
@@ -1776,30 +1715,34 @@ void CompositionalMultiphaseWell::implicitStepComplete( real64 const & GEOSX_UNU
       {
-        wellElemPressure[iwelem] += dWellElemPressure[iwelem];
+        wellElemPressureOld[iwelem] = wellElemPressure[iwelem];
+        connRateOld[iwelem] = connRate[iwelem];
         for( integer ic = 0; ic < numComp; ++ic )
         {
-          wellElemGlobalCompDensity[iwelem][ic] += dWellElemGlobalCompDensity[iwelem][ic];
+          wellElemGlobalCompDensityOld[iwelem][ic] = wellElemGlobalCompDensity[iwelem][ic];
```
`setValues`?

francoishamon replied:
Done

klevzoff reviewed (approved):
Nice work!

francoishamon commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp:
```diff
@@ -812,7 +800,7 @@ void SinglePhaseBase::applyDirichletBC( real64 const time_n,
                                                                    [=] GEOSX_HOST_DEVICE ( localIndex const a )
       {
-        return pres[a] + dPres[a];
+        return pres[a];
```
Cool, thanks for the suggestion, I did not know that this was possible

francoishamon commented on src/coreComponents/physicsSolvers/fluidFlow/wells/CompositionalMultiphaseWell.cpp:
```diff
@@ -1552,21 +1508,28 @@ void CompositionalMultiphaseWell::resetStateToBeginningOfStep( DomainPartition &
         for( integer ic = 0; ic < numComp; ++ic )
         {
-          dWellElemGlobalCompDensity[iwelem][ic] = 0;
+          wellElemGlobalCompDensity[iwelem][ic] = wellElemGlobalCompDensityOld[iwelem][ic];
```
Done

francoishamon commented on src/coreComponents/physicsSolvers/fluidFlow/wells/SinglePhaseWell.cpp:
```diff
@@ -890,41 +872,24 @@ void SinglePhaseWell::implicitStepSetup( real64 const & time,
-        wellElemPressure[iwelem] += dWellElemPressure[iwelem];
-        connRate[iwelem]         += dConnRate[iwelem];
+        wellElemPressureOld[iwelem] = wellElemPressure[iwelem];
+        connRateOld[iwelem]         = connRate[iwelem];
```
Done

francoishamon commented on src/coreComponents/physicsSolvers/fluidFlow/wells/SinglePhaseWell.cpp:
```diff
@@ -832,15 +809,20 @@ void SinglePhaseWell::resetStateToBeginningOfStep( DomainPartition & domain )
-        dWellElemPressure[iwelem] = 0;
-        dConnRate[iwelem] = 0;
+        wellElemPressure[iwelem] = wellElemPressureOld[iwelem];
+        connRate[iwelem] = connRateOld[iwelem];
```
Done

francoishamon commented on src/coreComponents/physicsSolvers/fluidFlow/proppantTransport/ProppantTransport.cpp:
```diff
@@ -469,33 +454,50 @@ void ProppantTransport::implicitStepComplete( real64 const & GEOSX_UNUSED_PARAM(
+        for( integer c = 0; c < numComponents; ++c )
         {
-          componentConc[ei][c] += dComponentConc[ei][c];
+          componentConcOld[ei][c] = componentConc[ei][c];
```
Done

francoishamon commented on src/coreComponents/physicsSolvers/fluidFlow/proppantTransport/ProppantTransport.cpp:
```diff
@@ -960,16 +955,21 @@ void ProppantTransport::resetStateToBeginningOfStep( DomainPartition & domain )
         for( localIndex c = 0; c < numComponents; ++c )
         {
-          dComponentConc[ei][c] = 0.0;
+          componentConc[ei][c] = componentConcOld[ei][c];
```
Done

francoishamon commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp:
```diff
@@ -256,44 +256,18 @@ void CompositionalMultiphaseHybridFVM::implicitStepSetup( real64 const & time_n,
     forAll< parallelDevicePolicy<> >( faceManager.size(), [=] GEOSX_HOST_DEVICE ( localIndex const iface )
     {
-      facePres[iface] += dFacePres[iface];
+      facePresOld[iface] = facePres[iface];
```
Done

francoishamon commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp:
```diff
@@ -926,11 +837,15 @@ void CompositionalMultiphaseHybridFVM::resetStateToBeginningOfStep( DomainPartit
+    forAll< parallelDevicePolicy<> >( faceManager.size(), [=] GEOSX_HOST_DEVICE ( localIndex const iface )
+    {
+      facePres[iface] = facePresOld[iface];
+    } );
```
Done

francoishamon commented on src/coreComponents/physicsSolvers/fluidFlow/proppantTransport/ProppantTransport.cpp:
```diff
@@ -725,7 +722,7 @@ void ProppantTransport::applyBoundaryConditions( real64 const time_n,
                                                                    [=] GEOSX_HOST_DEVICE ( localIndex const a )
       {
-        return proppantConc[a] + dProppantConc[a];
+        return proppantConc[a];
```
Done

francoishamon commented on src/coreComponents/physicsSolvers/fluidFlow/wells/WellSolverBase.cpp:
```diff
@@ -166,6 +163,13 @@ void WellSolverBase::implicitStepSetup( real64 const & time_n,
+{
+  GEOSX_UNUSED_VAR( time_n, dt, domain );
+}
+
```
Yes, for now I moved this function to the hpp file to have something more concise:
https://github.com/GEOSX/GEOSX/blob/26b65c4e44ff73cd0f44f341da402f50e5f669f0/src/coreComponents/physicsSolvers/fluidFlow/wells/WellSolverBase.hpp#L151-L153

If no one has objections, I could also change directly the function in `SolverBase` to:
```
void SolverBase::implicitStepComplete( real64 const & GEOSX_UNUSED_PARAM( time ),
                                       real64 const & GEOSX_UNUSED_PARAM( dt ),
                                       DomainPartition & GEOSX_UNUSED_PARAM( domain ) )
{}
```
which would allow me to entirely remove `WellSolverBase::implicitStepComplete`

francoishamon commented:
> Not expecting to be done so fast, but I was going to suggest adding specificity to the naming. So something like `pressure_n` and `pressure_np1`. With the `pressure/pressureOld` nomenclature, what exactly is meant by `Old` is a matter of agreed upon convention. What if, for instance, we wanted to have a method that required the value at the previous beginning of step, or some sort of multi-step method? I know this work is already essentially done...but it is good to think about if we want future flexibility.

@rrsettgast I just repost here my slack message, and I don't mind changing the content of the current PR before merging

Yes, I agree with your suggestion for `fieldOld` that could be a `field_n`, or anything that would explicitly say that the field is at the previous time level. That would be useful when we start to lag in iteration, as it would be a way to make a clear distinction between the lagged value and the value at the previous time level. Also useful for fractional step schemes, I agree with you. I don't mind doing this change in the current PR if everybody is on board with the change.

For `field_np1`, I am not a big fan, because it appears everywhere, and to be consistent, we should also have all the constitutive properties in the same format, like `density_np1`, etc, which would be overwhelming in my opinion. So I would prefer to stick with `field`, with the convention that if nothing is specified, the value is at the most recent iteration of the most recent time step

joshua-white commented:
I am happy with whatever nomenclature we adopt, but one advantage of `fieldOld` is that it is immediately clear what it means, while `field_n` might require documentation (some people use `n` to mean current timestep, others use `n+1` as the current timestep).  I agree for multistep methods it may be awkward to say `fieldOldOld`, but we might instead store data as an array to have `fieldOld[0], fieldOld[1]`, etc.

francoishamon commented:
I implemented the change from `fieldOld` to `field_n` in the flow solvers. To make everything uniform, I also renamed the "old" constitutive fields that appear in the flow solvers (multifluid, porosity, solid internal energy) and their corresponding accessors. To be specific, there is now a `phaseComponentFraction_n`, `porosity_n`, `solidInternalEnergy_n`.
